### PR TITLE
Update config file specification format

### PR DESCRIPTION
**Pull request by dragonejt**
# Description
* Updates the config `.ini` file specification format used by OnAIR to follow the requirements of #101 
* Updates execution engine config file parser to read correct keys
* Updates execution enginer unit tests to provide valid fake data

# Testing
* `python driver.py` ran with no errors
* `python driver.py -t` (all unit tests) succeeded

# Considerations
* I had used VSCode's autoformatter to format `test_execution_engine.py` when updating the fake data to verify the fake data was nested correctly. If we want to go back to the original format for some reason I can figure something out to revert the formatting, but as an action item I can also take a look at adding a linter and autoformatter if we want to ensure a consistent code style.
* I had posted a comment in #101 about it being difficult to find any references for the `.ini` file specification. Has it been considered to move to a more modern/commonly used python configuration file format such as [TOML](https://toml.io/en/) (which visually appears very similar to `.ini`)? Python has a built-in library for reading TOML files as well, and modern python libraries use a `pyproject.toml` file for configurations.
* **EDIT:** I have not been able to make the `OPTIONS` section optional, will modify functionality in the next commit

### CODE REVIEW

**Comment by dragonejt** [comment]
@the-other-james @asgibson just in case y'all haven't seen this PR, I updated the config file specification format to what was wanted in #128

**Comment by asgibson** [comment]
@dragonejt at a glance your coding looks very good, but my formal review will take a little more time. After my review, it will need a secondary review before a merge. There are two reviews ahead in the pipeline #125 and #127 (#98 is on hold).

In your comment, you mention #128, but as this comment is in 128, I assume you mean the issue #101. In my cursory review you appear to be hitting all the marks on the requests in that issue. While awaiting the formal review, are you able to create a feature branch for #101 and rebase it to branch [126](https://github.com/nasa/OnAIR/tree/126-provide-some-backwards-compatibility-with-previous-python-versions)? We like to use branches for issue tracking and the rebase will make merging smoother after the other PRs are complete.

Thank you very much for your time and effort, it is appreciated!

**Comment by dragonejt** [comment]
Sorry, yes I meant that this PR addresses #101 . Unfortunately, as I am not a collaborator on this repository I cannot create new branches on the repo or push to any branch on the repo. Instead, I forked the repository and submitted the PR from the fork to the original, as that was the guidance from [reading this](https://github.com/nasa/instructions/blob/master/docs/INSTRUCTIONS.md). There is the option  to be added as a collaborator to the repository on a case-by-case basis if that is something that we want to pursue. Otherwise, the best I can do is currently what is being done, which is forking the repo and submitting the PR from my fork onto the original repository.

**Comment by asgibson** [comment]
No worries, I thought that may be the case. Thank you for looking into it.

**Review by the-other-james — CHANGES_REQUESTED** [review]
Great work! And thank you for your patience with us!

Before I approve, please update the branch you are merging in to have the latest from our main. I think there are likely conflicts with the redis .ini file since Alan had made updates.

Also, please address my annoying nitpicks about formatting :)

Thanks!

File: onair/src/run_scripts/execution_engine.py
Comment (the-other-james) [reply onair/src/run_scripts/execution_engine.py:60]: Nevermind... leave the line breaks as you have them: Alan pointed out that https://peps.python.org/pep-0008/#maximum-line-length has the max as 79 characters.

File: onair/src/run_scripts/execution_engine.py
Comment (asgibson) [inline onair/src/run_scripts/execution_engine.py:106]: f-string does not allow splitting lines < Python 3.12
We are currently supporting Python 3.8-3.12. 
Suggest we allow the long lines or we'll need to end strings with \ (as well as not using the / within the braces {}) AND begin the next line left justified.
```
        new_message = f"Config file: '\
{config_filepath}', missing key: {e.args[0]}"
```

File: onair/src/run_scripts/execution_engine.py
Comment (asgibson) [inline onair/src/run_scripts/execution_engine.py:116]: f-string does not allow splitting lines < Python 3.12

File: onair/src/run_scripts/execution_engine.py
Comment (asgibson) [inline onair/src/run_scripts/execution_engine.py:121]: f-string does not allow splitting lines < Python 3.12

File: test/onair/src/run_scripts/test_execution_engine.py
Comment (asgibson) [inline test/onair/src/run_scripts/test_execution_engine.py:124]: f-string does not allow splitting lines < Python 3.12

File: test/onair/src/run_scripts/test_execution_engine.py
Comment (asgibson) [inline test/onair/src/run_scripts/test_execution_engine.py:406]: f-string does not allow splitting lines < Python 3.12

File: test/onair/src/run_scripts/test_execution_engine.py
Comment (asgibson) [inline test/onair/src/run_scripts/test_execution_engine.py:446]: f-string does not allow splitting lines < Python 3.12

File: test/onair/src/run_scripts/test_execution_engine.py
Comment (asgibson) [inline test/onair/src/run_scripts/test_execution_engine.py:499]: f-string does not allow splitting lines < Python 3.12

**Review by asgibson — CHANGES_REQUESTED** [review]
Good work! All 8 of the requested changes in #101 are addressed.
Some minor changes to make, but overall looks great.

File: .gitignore
Comment (asgibson) [reply .gitignore:5]: That I do.

**Comment by dragonejt** [comment]
Sorry for the delay, I've been busy. I'll work on these changes over the next few weeks and submit a revision.

**Comment by dragonejt** [comment]
Hey sorry I haven't gotten to revising this PR, I'm going to address the comments here over the next week or so. Are there any updates to how we use the .ini file that need to be reflected in this PR? Also, looking at the formatting changes requested, I was thinking about bringing in a Python code formatter so that I didn't need to manually format the codebase, such as [black](https://github.com/psf/black) or [Ruff](https://docs.astral.sh/ruff/). Is there any reason I should not use an autoformatter?

**Comment by dragonejt** [comment]
## Rev 2
* Remove `Dev_enabled` and `Viz_enabled` feature flags
* Format with `black` to achieve a consistent code style, future code can be formatted with `black .`. This also removes the multiline f-strings.
* Add all items in GitHub's default python .gitignore with a newline at the bottom. This covers all files previously ignored, and more files that we should probably ignore.
* Updated a unit test failure by changing how a class instance was created, please let me know if we need to change it back
```
FAILED test/onair/src/run_scripts/test_execution_engine.py::test_ExecutionEngine_parse_configs_skips_OPTIONS_when_the_required_section_OPTIONS_is_not_in_config - AttributeError: 'ExecutionEngine' object has no attribute 'IO_Enabled'
```
I believe `__new__` doesn't actually call `__init__`, so changing it to how normal class instances are created calls `__init__`.

Most of the changes are in `execution_engine.py` and `test_execution_engine.py`, the rest of the files changed should all be formatting. If we want to remove code formatting changes, I can remove this commit and just do one without the formatting.

## Testing
All Unit Tests succeeded.

**Comment by the-other-james** [comment]
Could you please remove the changes related to formatting? I don't want to mix formatting changes in with your other changes: it makes for a huge pull request (62 changed files) which obfuscates the config file update.

Reformatting of the entire code base should be done as its own pull request. Alan started looking into this a while ago: https://github.com/nasa/OnAIR/pull/98

**Comment by dragonejt** [comment]
A no-formatting revision has been pushed.

**Comment by dragonejt** [comment]
Just in case you two haven't seen this, @the-other-james @asgibson

**Review by the-other-james — APPROVED** [review]
Looks good to me! Tested default config and unit tests on my machine.

**Review by the-other-james — CHANGES_REQUESTED** [review]
I did a bit more testing and noticed that the kalman and redis configs hadn't been fully updated. I'm not sure if I can directly push my fixes to your merge request, so here's the diff that got things working on my end:

```
diff --git a/onair/config/kalman_csv_output_example.ini b/onair/config/kalman_csv_output_example.ini
index 06dab42..a1c927d 100644
--- a/onair/config/kalman_csv_output_example.ini
+++ b/onair/config/kalman_csv_output_example.ini
@@ -1,14 +1,17 @@
-[DEFAULT]
-TelemetryDataFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
+[FILES]
+TelemetryFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
 TelemetryFile = 700_crash_to_earth_1.csv
-TelemetryMetadataFilePath = onair/data/telemetry_configs/
+MetaFilePath = onair/data/telemetry_configs/
 MetaFile = data_physics_generation_CONFIG.json
-ParserFileName = onair/data_handling/csv_parser.py

+[DATA_HANDLING]
+DataSourceFile = onair/data_handling/csv_parser.py
+
+[PLUGINS]
 KnowledgeRepPluginDict = {'Kalman Filter': 'plugins/kalman'}
 LearnersPluginDict = {'csv output':'plugins/csv_output'}
 PlannersPluginDict = {}
 ComplexPluginDict = {}

-[RUN_FLAGS]
-IO_Flag = true
+[OPTIONS]
+IO_Enabled = true
diff --git a/onair/config/redis_example.ini b/onair/config/redis_example.ini
index b0a599b..3f70e42 100644
--- a/onair/config/redis_example.ini
+++ b/onair/config/redis_example.ini
@@ -3,7 +3,9 @@ TelemetryFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
 TelemetryFile = 700_crash_to_earth_1.csv
 MetaFilePath = onair/data/telemetry_configs/
 MetaFile = redis_example_CONFIG.json
-ParserFileName = onair/data_handling/redis_adapter.py
+
+[DATA_HANDLING]
+DataSourceFile = onair/data_handling/redis_adapter.py

 [PLUGINS]
 KnowledgeRepPluginDict = {'knowledge':'plugins/generic/__init__.py'}
```

**Comment by dragonejt** [comment]
> I did a bit more testing and noticed that the kalman and redis configs hadn't been fully updated. I'm not sure if I can directly push my fixes to your merge request, so here's the diff that got things working on my end:
> 
> ```
> diff --git a/onair/config/kalman_csv_output_example.ini b/onair/config/kalman_csv_output_example.ini
> index 06dab42..a1c927d 100644
> --- a/onair/config/kalman_csv_output_example.ini
> +++ b/onair/config/kalman_csv_output_example.ini
> @@ -1,14 +1,17 @@
> -[DEFAULT]
> -TelemetryDataFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
> +[FILES]
> +TelemetryFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
>  TelemetryFile = 700_crash_to_earth_1.csv
> -TelemetryMetadataFilePath = onair/data/telemetry_configs/
> +MetaFilePath = onair/data/telemetry_configs/
>  MetaFile = data_physics_generation_CONFIG.json
> -ParserFileName = onair/data_handling/csv_parser.py
> 
> +[DATA_HANDLING]
> +DataSourceFile = onair/data_handling/csv_parser.py
> +
> +[PLUGINS]
>  KnowledgeRepPluginDict = {'Kalman Filter': 'plugins/kalman'}
>  LearnersPluginDict = {'csv output':'plugins/csv_output'}
>  PlannersPluginDict = {}
>  ComplexPluginDict = {}
> 
> -[RUN_FLAGS]
> -IO_Flag = true
> +[OPTIONS]
> +IO_Enabled = true
> diff --git a/onair/config/redis_example.ini b/onair/config/redis_example.ini
> index b0a599b..3f70e42 100644
> --- a/onair/config/redis_example.ini
> +++ b/onair/config/redis_example.ini
> @@ -3,7 +3,9 @@ TelemetryFilePath = onair/data/raw_telemetry_data/data_physics_generation/Errors
>  TelemetryFile = 700_crash_to_earth_1.csv
>  MetaFilePath = onair/data/telemetry_configs/
>  MetaFile = redis_example_CONFIG.json
> -ParserFileName = onair/data_handling/redis_adapter.py
> +
> +[DATA_HANDLING]
> +DataSourceFile = onair/data_handling/redis_adapter.py
> 
>  [PLUGINS]
>  KnowledgeRepPluginDict = {'knowledge':'plugins/generic/__init__.py'}
> ```

Ah, I may have not fully changed the other configs, sorry about that. I'll get that fixed

**Comment by dragonejt** [comment]
@the-other-james It looks like what happened was that I did indeed miss the error in `redis_example.ini`, but the commit that added `onair/config/kalman_csv_output_example.ini` took place after I created my fork from the main repo, so my fork didn't have that file. The new commit I just pushed has the updates, and should work now.

**Review by asgibson — CHANGES_REQUESTED** [review]
Please revert the ExecutionEngine instantiations in the unit tests. By only allocating space for the instances it makes the tests more maintainable because changes in the `__init__` function will not affect the tests (as it will not run). This also requires the tests to explicitly set any attributes within the test itself because initialization was not performed on the instance, which provides more information about how the code under test (cut) operates.

File: test/onair/src/run_scripts/test_execution_engine.py
Comment (dragonejt) [reply test/onair/src/run_scripts/test_execution_engine.py]: I didn't realize that some of the tests were testing `__init__`, this has since been reverted and all instantiations should be calling `__new__`.

File: .gitignore
Comment (dragonejt) [reply .gitignore:1]: This .gitignore is the GitHub default gitignore, and is generated in any new Python GitHub repository: https://github.com/github/gitignore/blob/main/Python.gitignore

I noticed that `.DS_Store` for macOS was missing from the changed `.gitignore`, so I re-added that.

**Comment by dragonejt** [comment]
@asgibson the ExecutionEngine issue has been fixed, and I left a comment about where I got my Python .gitignore from. Please take a look when you have time.

**Review by asgibson — APPROVED** [review]

**Review by the-other-james — APPROVED** [review]
Great work! Thank you for you patience!

**Comment by dragonejt** [comment]
Hello, I don't have permissions to merge into the main branch, so one of you will need to merge whenever we're ready to merge.

**Pull request merged**
